### PR TITLE
Minor refactoring and cleanup to BooleanQuery code

### PR DESCRIPTION
**Pull request by twosom**
### Description
some cleanups (typos, use Java17 feature , improving readability)

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java [hunk 1]
```diff
@@ -204,11 +204,11 @@ BooleanQuery rewriteNoScoring(IndexSearcher indexSearcher) throws IOException {
...
       Query query = clause.getQuery();
       Query rewritten = new ConstantScoreQuery(query).rewrite(indexSearcher);
-      if (rewritten instanceof ConstantScoreQuery) {
-        rewritten = ((ConstantScoreQuery) rewritten).getQuery();
+      if (rewritten instanceof ConstantScoreQuery constantScoreQuery) {
+        rewritten = constantScoreQuery.getQuery();
       }
       BooleanClause.Occur occur = clause.getOccur();
-      if (occur == Occur.SHOULD && keepShould == false) {
+      if (occur == Occur.SHOULD && !keepShould) {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java:211]: we use `== false` intentionally across the codebase for better readability, can you keep it and other places in this class too?

File: lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java [hunk 2]
```diff
@@ -346,21 +338,15 @@ public Query rewrite(IndexSearcher indexSearcher) throws IOException {
...
-        for (BooleanClause clause : clauses) {
-          if (clause.getOccur() != Occur.FILTER) {
-            builder.add(clause);
-          }
-        }
-        for (Query filter : filters) {
-          builder.add(filter, Occur.FILTER);
-        }
+        clauses.stream().filter(clause -> clause.getOccur() != Occur.FILTER).forEach(builder::add);
+        filters.forEach(filter -> builder.add(filter, Occur.FILTER));
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java:349]: I think using streams above to compute `clauseCount` is fine as it makes the code less verbose while remaining easy to read. I'm less sure about this change: it's shorter, but it also reads harder for me.

File: lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java [hunk 3]
```diff
@@ -408,19 +393,15 @@ public Query rewrite(IndexSearcher indexSearcher) throws IOException {
...
-        }
+        shouldClauses.forEach(
+            (query, value) -> {
+              float boost = value.floatValue();
+              if (boost != 1f) {
+                query = new BoostQuery(query, boost);
+              }
+              builder.add(query, Occur.SHOULD);
+            });
+        clauses.stream().filter(clause -> clause.getOccur() != Occur.SHOULD).forEach(builder::add);
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java:404]: Likewise here, streams don't make the code easier to read?

**Review by jpountz — COMMENTED** [review]
There are some bits I like, such as leveraging Java 17's switch statements and instanceof sugar, and other bits I like less, such as replacing `x == false` with `!x` or abusing streams.

For the record, Lucene's 9.x branch which is still active compiles against Java 11, which is why you're not seeing much usage of Java 17 features on the main branch.

File: lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java (hunk 1 above)
Comment (twosom) [reply lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java:211]: Thanks to reply.
I'll keep `== false`😊

File: lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java (hunk 2 above)
Comment (twosom) [reply lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java:349]: What do you think about change this streams to like  `caluses.stream().filter(BooleanClause::isNotFilter).forEach(builder::add);`

Comment (twosom) [reply lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java:349]: and method isNotFilter will be like

```
public boolean isNotFilter() {
    return occur == Occur.FILTER;
}
```

File: lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java (hunk 3 above)
Comment (twosom) [reply lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java:404]: and this stream too.

`clauses.stream().filter(BooleanClause::isNotShould).forEach(builder::add);`

**Pull request closed without merging**